### PR TITLE
Rule for gradient accumulation

### LINKED ISSUE #130: Interface for gradient accumulation
### Motivation and description

Gradient accumulation is an important technique when the model does not fit in the memory with large batch size.

### Possible Implementation

Suggested by @ToucheSir, 

> defining an `AbstractRule` for accumulation which just negates the gradient. Then when when update! is called, https://github.com/FluxML/Optimisers.jl/blob/master/src/interface.jl#L98 will cancel out the negative. To make this work, run setup as usual on the model using this grad accum rule, then use that tree to update!(accum_tree, accum_grads, new_grads)  instead. Once you've finished accumulating, just run update!(opt_state_tree, model, accum_grads) as you would without accumulation.

### PR DESCRIPTION
**Pull request by CarloLucibello**
Fix #130 

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

File: src/rules.jl [hunk 1]
```diff
@@ -667,3 +667,39 @@ end
...
+function init(o::AccumGrad, x)
+  return (zero(x), 0, init(o.opt, x))
+end
+
+function apply!(o::AccumGrad, state, x, dx)
+  accum_dx, counter, opt_state = state
+  counter += 1
+  @.. accum_dx = accum_dx + dx
+  if counter == o.n
+    opt_state, dx′ = apply!(o.opt, opt_state, x, accum_dx ./ o.n)
```
Comment (CarloLucibello) [inline src/rules.jl:698]: A first version was
```suggestion
    @.. accum_dx = accum_dx / o.n
    opt_state, dx′ = apply!(o.opt, opt_state, x, accum_dx)
    @.. accum_dx = zero(accum_dx)
```
but the zeroing in the last line kills the `@lazy` gradient computed by `o.opt`, so it looks like we need to allocate a new array `accum_dx ./ o.n` but it would be nice to avoid that.

File: test/rules.jl [hunk 2]
```diff
@@ -244,3 +244,27 @@ VERSION < v"1.9-" && @testset "using Yota" begin
...
+  x0 = rand(5)
+  x = copy(x0)
+  lr = 0.01
+  tree = Optimisers.setup(AccumGrad(Descent(lr), 3), x)
+
+  g1 = rand(5)
+  tree, x1 = Optimisers.update(tree, x, g1)
+  @test x1 ≈ x
+  @test x1 ≈ x0 
+  @test_broken x1 === x
```
Comment (CarloLucibello) [inline test/rules.jl:258]: Why this is not passing? Can we avoid making a copy?

Comment (darsnack) [reply test/rules.jl:258]: `update` always copies first. `update!` will not copy if the underlying storage is mutable.

File: src/rules.jl [hunk 3]
```diff
@@ -667,3 +667,39 @@ end
...
+  
+  function AccumGrad(opt::R, n::Int) where R<:AbstractRule
+    n > 0 || throw(ArgumentError("AccumGrad must accumulate at least one gradient"))
+    return new{R}(opt, n)  
+  end
+end
+
+function init(o::AccumGrad, x)
+  return (zero(x), 0, init(o.opt, x))
+end
```
Comment (darsnack) [inline src/rules.jl:691]: This will need a custom `adjust` method

File: src/interface.jl [hunk 4]
```diff
@@ -96,6 +96,7 @@ function _update!(ℓ::Leaf, x; grads, params)
 end
 
 subtract!(x, x̄) = maywrite(x) ? (x .= x .- x̄) : eltype(x).(x .- x̄)
+subtract!(x, x̄::Nothing) = x
```
Comment (mcabbott) [inline src/interface.jl:99]:
```suggestion
subtract!(x, x̄::Zero) = x
```

File: src/rules.jl (hunk 1 above)
Comment (mcabbott) [reply src/rules.jl:698]: Maybe you can zero it at the start when count==0, rather than at the end?

File: test/rules.jl [hunk 5]
```diff
@@ -244,3 +244,27 @@ VERSION < v"1.9-" && @testset "using Yota" begin
     @test loss(w, w′) < 0.001
   end
 end
+
+@testset "AccumGrad" begin
+  x0 = rand(5)
+  x = copy(x0)
+  lr = 0.01
+  tree = Optimisers.setup(AccumGrad(Descent(lr), 3), x)
```
Comment (mcabbott) [inline test/rules.jl:252]: A torture test would use something with a large momentum, so that it fails if the rule is applied 3 times, rather that once for the total gradient. But not sure that's necessary.

File: test/rules.jl [hunk 6]
```diff
@@ -244,3 +244,27 @@ VERSION < v"1.9-" && @testset "using Yota" begin
...
+  @test x2 ≈ x
+  @test x2 ≈ x0 
+  g3 = rand(5)
+  tree, x3 = Optimisers.update(tree, x2, g3)
+  @test x3 ≈ x0 .- lr .* (g1 .+ g2 .+ g3) ./ 3
+  g4 = rand(5)
+  
+  tree, x4 = Optimisers.update(tree, x3, g4)
+  @test x4 ≈ x3
+end
```
Comment (mcabbott) [inline test/rules.jl:270]: In rules.jl, adding some cases to `RULES = [` will run tests of this thing with StaticArrays, complex numbers, etc.

File: src/rules.jl [hunk 7]
```diff
@@ -667,3 +667,39 @@ end
...
+"""
+  AccumGrad(opt::AbstractRule, n::Int)
+
+A wrapper for an optimisation rule which accumulates gradients over `n` steps before
+applying the rule. This is useful for training with batch sizes larger than the
+available memory.
+"""
+struct AccumGrad{R<:AbstractRule} <: AbstractRule
+  opt::R
+  n::Int
```
Comment (mcabbott) [inline src/rules.jl:681]: The big question is whether the design should be 
```
AccumGrad(OptimiserChain(ClipGrad(), AdamW()), 3)
```
or
```
OptimiserChain(AccumGrad(3), ClipGrad(), AdamW())
```

I haven't thought too hard yet, but was the latter considered somewhere? `AccumGrad(3) => ClipGrad() => AdamW()`

Comment (CarloLucibello) [reply src/rules.jl:681]: > I haven't thought too hard yet, but was the latter considered somewhere?

briefly in #130, but now that I've digged a bit into it I like `OptimiserChain(AccumGrad(3), ClipGrad(), AdamW())`
more, it's nicer to read, more decoupled, and one make sense also of different permutations of the arguments.

We could just tweak the `OptimiserChain` to stop the propagation when it hits `nothing` signal. 
Or if we think that propagating `nothing` could also make sense, we could define a `NoUpdate` type.

File: src/interface.jl [hunk 8]
```diff
@@ -1,8 +1,20 @@
...
+
+A dummy gradient that will not be applied to the parameters.
+Can be returned by `apply!` rules to indicate that no update should be done.
+
+When returned by a rule in a [`OptimiserChain`](@ref), it will also prevent the 
+following rules from being called.
+"""
+struct NoUpdate end
+
+const Zero = Union{Nothing, AbstractZero, NoUpdate}  # Union{Zygote, Diffractor, Optimisers}
```
Comment (mcabbott) [inline src/interface.jl:17]:
```suggestion
const Zero = Union{Nothing, AbstractZero}  # Union{Zygote, Diffractor}
```

File: src/rules.jl [hunk 9]
```diff
@@ -654,11 +658,15 @@ init(o::OptimiserChain, x::AbstractArray) = map(opt -> init(opt, x), o.opts)
...
   foldl(tuple.(o.opts, states); init = ((), dx)) do (states′, dx′), (opt, state)
-    state′, dx′ = apply!(opt, state, x, dx′, dxs...)
+    state′, dx′ = _apply_in_chain!(opt, state, x, dx′, dxs...)
     return (states′..., state′), dx′
   end
 end
 
+_apply_in_chain!(opt, state, x, dx′, dxs...) = apply!(opt, state, x, dx′, dxs...)
+_apply_in_chain!(opt, state, x, dx′::NoUpdate, dxs...) = state, NoUpdate()
+
```
[the remark this answers is not in the document]
Comment (CarloLucibello) [reply src/rules.jl:668]: I agree that using a `Zero` as marker is consistent with what currently do, so maybe we should go with it in this PR. 
But what we do seems semantically wrong. Isn't `nothing` supposed to be a placeholder for `zero(x)`. Shouldn't
these two produce the same effects:
```julia
Optimisers.update!(st, x, nothing)
Optimisers.update!(st, x, zero(x))
```
?

[the remark this answers is not in the document]
Comment (mcabbott) [reply src/rules.jl:668]: I suspect this is more an accident than a design, but I'm not sure it's an awful one. 

If you are doing ordinary AD and happen to get an array of zeros on some batch, probably you do want that to update the momenta etc. 

But you won't get `nothing` just because of the data in that batch. Instead, you'll get it because you are e.g. doing transfer learning, or the generator & discriminator on even/odd steps, or something like that. You will get `nothing`  not for one array, but for a whole part of the model. And it seems like you probably don't want to update the momenta for the part of the model not being trained, but instead just ignore them completely.

[the remark this answers is not in the document]
Comment (CarloLucibello) [reply src/rules.jl:668]: I think your examples should correspond to the opt_tree only having part of the model or to using different trees for discriminator and generator. Anyways, it's better to continue this discussion somewhere else, I'll open an issue.

**Review by mcabbott — APPROVED** [review]

File: src/rules.jl [hunk 10]
```diff
@@ -667,3 +672,60 @@ end
 
 adjust(ℓ::OptimiserChain, eta::Real) = OptimiserChain(map(opt -> adjust(opt, eta), ℓ.opts)...)
 adjust(ℓ::OptimiserChain; kw...) = OptimiserChain(map(opt -> adjust(opt; kw...), ℓ.opts)...)
+
+
+"""
+  AccumGrad(n::Int)
```
Comment (darsnack) [inline src/rules.jl:678]: Minor typo for rendering
```suggestion
    AccumGrad(n::Int)
```
